tests/test_binderbot.py

```diff
@@ -4,67 +4,160 @@
     return nbformat.from_dict(nbdata)
 
 
-def test_cli_upload_execute_download(tmp_path, example_nb_data):
-    """Test the CLI."""
+@pytest.fixture()
+def binder_url():
+    return 'https://mybinder.org'
+
+@pytest.mark.asyncio
+async def test_binder_start_stop(binder_url):
+    """
+    Test that our binder starts and stops
+    """
+    async with binderbot.BinderUser(binder_url, 'binder-examples/requirements', 'master') as jovyan:
+        await jovyan.start_binder()
+        headers = {'Authorization': f'token {jovyan.token}'}
+        resp = await jovyan.session.get(
+            jovyan.notebook_url / 'api/status',
+            headers=headers
+        )
+        assert resp.status == 200
+
+        await jovyan.shutdown_binder()
+        resp = await jovyan.session.get(
+            jovyan.notebook_url / 'api/status',
+            headers=headers
+        )
+
+        assert resp.status == 503
+
+
+@pytest.mark.asyncio
+async def test_nbclient_run_code(local_notebook: LocalNotebookServer):
+    log = structlog.get_logger().bind()
+    async with aiohttp.ClientSession() as session:
+        nbclient = binderbot.NotebookClient(session, local_notebook.url, local_notebook.token, log)
+
+        async with nbclient.start_kernel() as kernel:
+            stdout, stderr = await kernel.run_code(f"""
+            print('hi')
+            """)
+
+        assert stderr.strip() == ""
+        assert stdout.strip() == 'hi'
 
-    os.chdir(tmp_path)
-    fname = "example_notebook.ipynb"
-    with open(fname, 'w', encoding='utf-8') as f:
-        nbformat.write(example_nb_data, f)
 
-    env = {"MY_VAR": "SECRET"}
-    runner = CliRunner(env=env)
-    args = ["--binder-url", "http://mybinder.org",
-            "--repo", "binder-examples/requirements",
-            "--ref", "master", "--nb-timeout", "10",
-            "--pass-env-var",  "MY_VAR",
-            fname]
-    result = runner.invoke(cli.main, args)
-    assert result.exit_code == 0, result.output
+@pytest.mark.asyncio
+async def test_upload(local_notebook: LocalNotebookServer):
+    log = structlog.get_logger().bind()
+    async with aiohttp.ClientSession() as session:
+        nbclient = binderbot.NotebookClient(session, local_notebook.url, local_notebook.token, log)
 
-    with open(fname) as f:
-        nb = nbformat.read(f, as_version=4)
+        fname = "example-notebook.ipynb"
+        filepath = local_notebook.cwd / fname
+        input_notebook = make_code_notebook(["print('hello')"])
+        with open(filepath, 'w', encoding='utf-8') as f:
+            nbformat.write(input_notebook, f)
 
-    hostname = nb['cells'][0]['outputs'][0]['text']
-    assert hostname.startswith('jupyter-binder-')
-    remote_env_var_value = nb['cells'][1]['outputs'][0]['text']
-    assert remote_env_var_value.rstrip() == env['MY_VAR']
+        async with nbclient.start_kernel() as kernel:
+            await kernel.execute_notebook(
+                fname,
+                timeout=60,
+                )
```

Comment:
I'm a little concerned here that the feature to pass environment variables is not being tested any more. It would be easy to cover this feature by adding `env_vars` to the call to `execute_notebook`.

Or maybe a separate test for that feature would be best.

Reply:
@rabernat good catch. I'll add another unit test for it.